internal/delete_handler.go

```diff
@@ -154,6 +155,84 @@ func FindTargetRetain(folder storage.Folder,
 	return findTarget(folder, greater, choiceFunc)
 }
 
+func FindTargetRetainAfterName(folder storage.Folder,
+	retentionCount int, name string, modifier int,
+	isFullBackup func(object storage.Object) bool,
+	greater func(object1, object2 storage.Object) bool) (storage.Object, error) {
+
+	less := func(object1, object2 storage.Object) bool {return greater(object2, object1)}
+
+	choiceFuncRetain := getRetainChoiceFunc(retentionCount, modifier, isFullBackup)
+	if choiceFuncRetain == nil {
+		return nil, utility.NewForbiddenActionError("Not allowed modifier for 'delete before'")
+	}
+	meetName := false
+	choiceFuncAfterName := func(object storage.Object) bool {
+		meetName = meetName || strings.HasPrefix(object.GetName(), name)
+		if modifier == NoDeleteModifier {
+			return meetName
+		} else {
+			return meetName && isFullBackup(object)
+		}
+	}
+	if choiceFuncAfterName == nil {
+		return nil, utility.NewForbiddenActionError("Not allowed modifier for 'delete before'")
+	}
+
+	target1, err := findTarget(folder, greater, choiceFuncRetain)
+	if err != nil {
+		return nil, err
+	}
+	target2, err := findTarget(folder, less, choiceFuncAfterName)
+	if err != nil {
+		return nil, err
+	}
+
+	if greater(target2, target1) {
+		return target1, nil
+	} else {
+		return target2, nil
+	}
+}
+
+func FindTargetRetainAfterTime(folder storage.Folder,
+	retentionCount int,
+	timeLine time.Time,
+	modifier int,
+	isFullBackup func(object storage.Object) bool,
+	greater func(object1, object2 storage.Object) bool) (storage.Object, error) {
+
+	less := func(object1, object2 storage.Object) bool {return greater(object2, object1)}
+
+	choiceFuncRetain := getRetainChoiceFunc(retentionCount, modifier, isFullBackup)
+	if choiceFuncRetain == nil {
+		return nil, utility.NewForbiddenActionError("Not allowed modifier for 'delete retain'")
+	}
+	choiceFuncAfter := func(object storage.Object) bool {
+		timeCheck := timeLine.Before(object.GetLastModified()) || timeLine.Equal(object.GetLastModified())
+		if modifier == NoDeleteModifier {
+			return timeCheck
+		} else {
```

Comment:
you can make this easier(remove else):

if cond {
     return val1
}
return val2